{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f94adbd4",
   "metadata": {},
   "source": [
    "# External Data Retrieval\n",
    "Uses external APIs for more data. Saves data for further analysis and offline use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bb17dc4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# All imports\n",
    "import pandas as pd\n",
    "import pickle\n",
    "from pathlib import Path\n",
    "import utils"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f7446a7",
   "metadata": {},
   "source": [
    "## Data retrieved from the Ensembl API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9d0e7f97",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load DataFrames\n",
    "train_df = pd.read_csv(\n",
    "    utils.RAW_TRAIN_PATH,\n",
    "    usecols=[\"accession\", \"scoreset\", \"ensp\", \"pos\", \"ref_long\", \"ref_short\", \"alt_long\", \"alt_short\"],\n",
    "    dtype={\"accession\": str, \"scoreset\": str, \"ensp\": str, \"pos\": int, \"ref_long\": str, \"ref_short\": str, \"alt_long\": str, \"alt_short\": str}\n",
    ")\n",
    "test_df = pd.read_csv(\n",
    "    utils.RAW_TEST_PATH,\n",
    "    usecols=[\"accession\", \"ensp\", \"pos\", \"ref_long\", \"ref_short\", \"alt_long\", \"alt_short\"],\n",
    "    dtype={\"accession\": str, \"ensp\": str, \"pos\": int, \"ref_long\": str, \"ref_short\": str, \"alt_long\": str, \"alt_short\": str}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1905948",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensembl Protein ID to Sequence mapping\n",
    "train_ensp_sequence_map: dict[str, str] = {}\n",
    "test_ensp_sequence_map: dict[str, str] = {}\n",
    "train_sequence_series: pd.Series = train_df[\"ensp\"].apply(lambda ensp: utils.get_full_sequence(ensp, train_ensp_sequence_map))\n",
    "test_sequence_series: pd.Series = test_df[\"ensp\"].apply(lambda ensp: utils.get_full_sequence(ensp, test_ensp_sequence_map))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e0de49d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the mappings using pickle\n",
    "utils.dict_to_pickle(train_ensp_sequence_map, utils.TRAIN_ENSP_SEQUENCE_MAP_PATH)\n",
    "utils.dict_to_pickle(test_ensp_sequence_map, utils.TEST_ENSP_SEQUENCE_MAP_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0add2a34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensembl Variant Effect Predictor (VEP)\n",
    "train_vep_data: list[dict] = utils.vep_from_pickle(utils.TRAIN_VEP_DATA_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c65f9e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "178554\n",
      "dict_keys(['seq_region_name', 'input', 'assembly_name', 'allele_string', 'strand', 'transcript_consequences', 'most_severe_consequence', 'end', 'start', 'colocated_variants', 'id'])\n",
      "frameshift_variant\n",
      "dict_keys(['cdna_end', 'consequence_terms', 'cdna_start', 'impact', 'strand', 'cds_start', 'amino_acids', 'transcript_id', 'hgnc_id', 'gene_symbol_source', 'cds_end', 'protein_start', 'variant_allele', 'gene_symbol', 'gene_id', 'protein_end', 'codons', 'biotype'])\n",
      "{'seq_region_name': 'X', 'input': 'X:g.15594962..15594964delinsTTA', 'assembly_name': 'GRCh38', 'allele_string': 'C/TTA', 'strand': 1, 'transcript_consequences': [{'cdna_end': 277, 'consequence_terms': ['frameshift_variant'], 'cdna_start': 277, 'impact': 'HIGH', 'strand': -1, 'cds_start': 228, 'amino_acids': 'Q/HX', 'transcript_id': 'ENST00000252519', 'hgnc_id': 'HGNC:13557', 'gene_symbol_source': 'HGNC', 'cds_end': 228, 'protein_start': 76, 'variant_allele': 'TTA', 'gene_symbol': 'ACE2', 'gene_id': 'ENSG00000130234', 'protein_end': 76, 'codons': 'caG/caTAA', 'biotype': 'protein_coding'}, {'codons': 'caG/caTAA', 'biotype': 'protein_coding', 'protein_end': 76, 'protein_start': 76, 'cds_end': 228, 'variant_allele': 'TTA', 'gene_symbol': 'ACE2', 'gene_id': 'ENSG00000130234', 'hgnc_id': 'HGNC:13557', 'gene_symbol_source': 'HGNC', 'amino_acids': 'Q/HX', 'transcript_id': 'ENST00000427411', 'impact': 'HIGH', 'strand': -1, 'cds_start': 228, 'cdna_start': 531, 'consequence_terms': ['frameshift_variant'], 'cdna_end': 531}, {'consequence_terms': ['frameshift_variant'], 'cdna_end': 277, 'cdna_start': 277, 'amino_acids': 'Q/HX', 'transcript_id': 'ENST00000678046', 'impact': 'HIGH', 'strand': -1, 'cds_start': 228, 'protein_start': 76, 'cds_end': 228, 'variant_allele': 'TTA', 'gene_symbol': 'ACE2', 'gene_id': 'ENSG00000130234', 'hgnc_id': 'HGNC:13557', 'gene_symbol_source': 'HGNC', 'codons': 'caG/caTAA', 'biotype': 'protein_coding', 'protein_end': 76}, {'gene_symbol_source': 'HGNC', 'hgnc_id': 'HGNC:13557', 'gene_symbol': 'ACE2', 'gene_id': 'ENSG00000130234', 'protein_start': 76, 'cds_end': 228, 'variant_allele': 'TTA', 'protein_end': 76, 'biotype': 'protein_coding', 'codons': 'caG/caTAA', 'cdna_end': 431, 'consequence_terms': ['frameshift_variant'], 'cdna_start': 431, 'strand': -1, 'cds_start': 228, 'impact': 'HIGH', 'transcript_id': 'ENST00000678073', 'amino_acids': 'Q/HX'}, {'protein_end': 76, 'biotype': 'protein_coding', 'codons': 'caG/caTAA', 'gene_symbol_source': 'HGNC', 'hgnc_id': 'HGNC:13557', 'gene_id': 'ENSG00000130234', 'gene_symbol': 'ACE2', 'variant_allele': 'TTA', 'cds_end': 228, 'protein_start': 76, 'cds_start': 228, 'strand': -1, 'impact': 'HIGH', 'transcript_id': 'ENST00000679162', 'amino_acids': 'Q/HX', 'consequence_terms': ['frameshift_variant'], 'cdna_end': 277, 'cdna_start': 277}, {'strand': -1, 'cds_start': 228, 'impact': 'HIGH', 'transcript_id': 'ENST00000679212', 'amino_acids': 'Q/HX', 'consequence_terms': ['frameshift_variant'], 'cdna_start': 277, 'cdna_end': 277, 'protein_end': 76, 'biotype': 'protein_coding', 'codons': 'caG/caTAA', 'gene_symbol_source': 'HGNC', 'hgnc_id': 'HGNC:13557', 'gene_symbol': 'ACE2', 'gene_id': 'ENSG00000130234', 'protein_start': 76, 'cds_end': 228, 'variant_allele': 'TTA'}, {'codons': 'caG/caTAA', 'biotype': 'protein_coding', 'protein_end': 76, 'variant_allele': 'TTA', 'protein_start': 76, 'cds_end': 228, 'gene_symbol': 'ACE2', 'gene_id': 'ENSG00000130234', 'hgnc_id': 'HGNC:13557', 'gene_symbol_source': 'HGNC', 'amino_acids': 'Q/HX', 'transcript_id': 'ENST00000679278', 'impact': 'HIGH', 'cds_start': 228, 'strand': -1, 'consequence_terms': ['frameshift_variant'], 'cdna_start': 277, 'cdna_end': 277}, {'amino_acids': 'Q/HX', 'transcript_id': 'ENST00000680121', 'impact': 'HIGH', 'strand': -1, 'cds_start': 228, 'consequence_terms': ['frameshift_variant'], 'cdna_end': 357, 'cdna_start': 357, 'codons': 'caG/caTAA', 'biotype': 'protein_coding', 'protein_end': 76, 'protein_start': 76, 'cds_end': 228, 'variant_allele': 'TTA', 'gene_symbol': 'ACE2', 'gene_id': 'ENSG00000130234', 'hgnc_id': 'HGNC:13557', 'gene_symbol_source': 'HGNC'}, {'impact': 'HIGH', 'cds_start': 228, 'strand': -1, 'amino_acids': 'Q/HX', 'transcript_id': 'ENST00000871928', 'consequence_terms': ['frameshift_variant'], 'cdna_start': 297, 'cdna_end': 297, 'protein_end': 76, 'codons': 'caG/caTAA', 'biotype': 'protein_coding', 'hgnc_id': 'HGNC:13557', 'gene_symbol_source': 'HGNC', 'variant_allele': 'TTA', 'protein_start': 76, 'cds_end': 228, 'gene_symbol': 'ACE2', 'gene_id': 'ENSG00000130234'}, {'biotype': 'protein_coding', 'codons': 'caG/caTAA', 'protein_end': 76, 'gene_symbol': 'ACE2', 'gene_id': 'ENSG00000130234', 'protein_start': 76, 'cds_end': 228, 'variant_allele': 'TTA', 'gene_symbol_source': 'HGNC', 'hgnc_id': 'HGNC:13557', 'transcript_id': 'ENST00000871929', 'amino_acids': 'Q/HX', 'strand': -1, 'cds_start': 228, 'impact': 'HIGH', 'consequence_terms': ['frameshift_variant'], 'cdna_end': 262, 'cdna_start': 262}, {'consequence_terms': ['frameshift_variant'], 'cdna_end': 262, 'cdna_start': 262, 'cds_start': 228, 'strand': -1, 'impact': 'HIGH', 'transcript_id': 'ENST00000871930', 'amino_acids': 'Q/HX', 'gene_symbol_source': 'HGNC', 'hgnc_id': 'HGNC:13557', 'gene_id': 'ENSG00000130234', 'gene_symbol': 'ACE2', 'variant_allele': 'TTA', 'protein_start': 76, 'cds_end': 228, 'protein_end': 76, 'biotype': 'protein_coding', 'codons': 'caG/caTAA'}, {'cds_start': 228, 'strand': -1, 'impact': 'HIGH', 'transcript_id': 'ENST00000871931', 'amino_acids': 'Q/HX', 'cdna_end': 262, 'consequence_terms': ['frameshift_variant'], 'cdna_start': 262, 'protein_end': 76, 'biotype': 'protein_coding', 'codons': 'caG/caTAA', 'gene_symbol_source': 'HGNC', 'hgnc_id': 'HGNC:13557', 'gene_symbol': 'ACE2', 'gene_id': 'ENSG00000130234', 'variant_allele': 'TTA', 'protein_start': 76, 'cds_end': 228}, {'amino_acids': 'Q/HX', 'transcript_id': 'ENST00000954330', 'impact': 'HIGH', 'cds_start': 228, 'strand': -1, 'cdna_start': 438, 'consequence_terms': ['frameshift_variant'], 'cdna_end': 438, 'codons': 'caG/caTAA', 'biotype': 'protein_coding', 'protein_end': 76, 'variant_allele': 'TTA', 'cds_end': 228, 'protein_start': 76, 'gene_symbol': 'ACE2', 'gene_id': 'ENSG00000130234', 'hgnc_id': 'HGNC:13557', 'gene_symbol_source': 'HGNC'}, {'cdna_end': 308, 'consequence_terms': ['frameshift_variant'], 'cdna_start': 308, 'impact': 'HIGH', 'cds_start': 228, 'strand': -1, 'amino_acids': 'Q/HX', 'transcript_id': 'ENST00000954331', 'hgnc_id': 'HGNC:13557', 'gene_symbol_source': 'HGNC', 'variant_allele': 'TTA', 'protein_start': 76, 'cds_end': 228, 'gene_symbol': 'ACE2', 'gene_id': 'ENSG00000130234', 'protein_end': 76, 'codons': 'caG/caTAA', 'biotype': 'protein_coding'}], 'most_severe_consequence': 'frameshift_variant', 'end': 15594962, 'start': 15594962, 'colocated_variants': [{'somatic': 1, 'phenotype_or_disease': 1, 'id': 'COSV99383008', 'end': 15594962, 'seq_region_name': 'X', 'var_synonyms': {'COSMIC': ['COSM8630049']}, 'allele_string': 'COSMIC_MUTATION', 'strand': 1, 'start': 15594962}], 'id': 'X:g.15594962..15594964delinsTTA'}\n",
      "{'frameshift_variant', 'inframe_insertion', 'stop_gained', 'splice_acceptor_variant', 'splice_donor_variant', '5_prime_UTR_variant'}\n"
     ]
    }
   ],
   "source": [
    "# Example usage of VEP data\n",
    "# NOTE: \"strand\" is always 1. To get actual strand, use [\"transcript_consequences\"][0][\"strand\"]\n",
    "print(len(train_vep_data))\n",
    "print(train_vep_data[0].keys())\n",
    "print(train_vep_data[0][\"most_severe_consequence\"])\n",
    "print(train_vep_data[0][\"transcript_consequences\"][0].keys())\n",
    "print(train_vep_data[0])\n",
    "# Print all unique variant types in the VEP data\n",
    "variant_types: set[str] = set()\n",
    "for row in train_vep_data:\n",
    "    variant_types.add(row.get(\"most_severe_consequence\", \"error\"))\n",
    "print(variant_types)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0591e7f0",
   "metadata": {},
   "source": [
    "## Data retrieved from the MaveDB API"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aasp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
